docs/adr/0009-remote-types-interfaces-and-proc-macros.md

```diff
@@ -0,0 +1,273 @@
+        self.is::<foo::Error>()
+    }
+
+    fn is_bar_error(&self) -> bool {
+        self.is::<foo::Error>()
+    }
+}
+```
+
+### [Option 1a] use a function-style macro
+
+The same idea could also be spelled out using a function-style macro rather than an attribute macro:
+
+```rust
+type LogLevel = log::Level;
+
+uniffi::remote!(
+    pub enum LogLevel {
+        Error = 1,
+        Warn = 2,
+        Info = 3,
+        Debug = 4,
+        Trace = 5,
+    }
+);
+
+type AnyhowError = anyhow::Error;
+
+uniffi::remote!(
+    impl AnyhowError {
+        pub fn to_string(&self) -> String;
+
+        fn is_foo_error(&self) -> bool {
+            self.is::<foo::Error>()
+        }
+
+        fn is_bar_error(&self) -> bool {
+            self.is::<foo::Error>()
+        }
+    }
+);
+```
+
+### [Option 2] use custom-type conversion to expose the type
+
+An alternate strategy would be to use a custom-type conversion from that type into a local type that does implement the UniFFI traits.
+These examples will use the custom type syntax from #2087, since I think it looks nicer than the current `UniffiCustomTypeConverter` based code.
+
+```rust
+/// Define a type that mirrors `Log::Level`
+#[derive(uniffi::Enum)]
+pub enum LogLevel {
+    Error = 1,
+    Warn = 2,
+    Info = 3,
+    Debug = 4,
+    Trace = 5,
+}
+
+/// Define a custom type conversion from `log::Level` to the above type.
+uniffi::custom_type!(log::Level, LogLevel, {
+  from_custom: |l| match l {
+    log::Level::Error => LogLevel::Error,
+    log::Level::Warn => LogLevel::Warn,
+    log::Level::Info => LogLevel::Info,
+    log::Level::Debug => LogLevel::Debug,
+    log::Level::Trace => LogLevel::Trace,
+  },
+  try_into_custom: |l| Ok(match l ({
+    LogLevel::Error => log::Level::Error,
+    LogLevel::Warn => log::Level::Warn,
+    LogLevel::Info => log::Level::Info,
+    LogLevel::Debug => log::Level::Debug,
+    LogLevel::Trace => log::Level::Trace,
+  })
+})
+
+/// Interfaces can use the newtype pattern
+#[derive(uniffi::Object)]
+pub struct AnyhowError(anyhow::Error);
+
+uniffi::custom_newtype!(anyhow::Error, AnyhowError).
+
+// We can define methods directly with this approach, no need for extension traits.
+#[uniffi::export]
+impl AnyhowError {
+    fn is_foo_error(&self) -> bool {
+        self.0.is::<foo::Error>()
+    }
+
+    fn is_bar_error(&self) -> bool {
+        self.0.is::<bar::Error>()
+    }
+
+    fn message(&self) -> String {
+        self.0.to_string()
+    }
+}
+```
+
+#### Two types
+
+One drawback of this approach is that we have to equivalent, but different types.
+Rust code would need to use `anyhow::Error` in their signatures, while foreign code would use `AnyhowError`.
+Since the types are almost exactly the same, but named slightly different and with slightly different methods, it can be awkward to document this distinction -- both by UniFFI for library authors and by library authors for their consumers.
+
+### [Option 3] hybrid approach
+
```

Comment:
I don't really understand what is being proposed here, so I can't say I prefer it - maybe this should either be fleshed out or removed?